synapse/lib/snap.py

```diff
@@ -55,6 +55,324 @@ def _isTagInc(self, tag):
+
+        # TODO can we mandate any subs are returned pre-normalized?
+        propsubs = norminfo.get('subs')
+        if propsubs is not None:
+            for subname, subvalu in propsubs.items():
+                full = f'{prop.name}:{subname}'
+                if self.form.props.get(full) is not None:
+                    await self.set(full, subvalu)
+
+        propadds = norminfo.get('adds')
+        if propadds is not None:
+            for addname, addvalu, addinfo in propadds:
+                await self.ctx.addNode(addname, addvalu, norminfo=addinfo)
+
+        return True
+
+class SnapEditor:
+    '''
+    A SnapEditor allows tracking node edits with subs/deps as a transaction.
+    '''
+    def __init__(self, snap):
+        self.snap = snap
+        self.protonodes = {}
+        self.maxnodes = snap.core.maxnodes
+
+    def getNodeEdits(self):
+        nodeedits = []
+        for protonode in self.protonodes.values():
+            nodeedit = protonode.getNodeEdit()
+            if nodeedit is not None:
+                nodeedits.append(nodeedit)
+        return nodeedits
+
+    async def addNode(self, formname, valu, props=None, norminfo=None):
+
+        self.snap.core._checkMaxNodes()
+
+        form = self.snap.core.model.form(formname)
+        if form is None:
+            mesg = f'No form named {formname}.'
+            return await self.snap._raiseOnStrict(s_exc.NoSuchForm, mesg)
+
+        if form.isrunt:
+            mesg = f'Cannot make runt nodes: {formname}.'
+            return await self.snap._raiseOnStrict(s_exc.IsRuntForm, mesg)
+
+        if form.locked:
+            mesg = f'Form {form.full} is locked due to deprecation.'
+            return await self.snap._raiseOnStrict(s_exc.IsDeprLocked, mesg)
+
+        if norminfo is None:
+            try:
+                valu, norminfo = form.type.norm(valu)
+            except s_exc.BadTypeValu as e:
+                if self.snap.strict: raise e
+                await self.snap.warn('addNode() BadTypeValu {formname}={valu} {e}')
+                return None
+
+        protonode = await self._initProtoNode(form, valu, norminfo)
+        if props is not None:
+            [await protonode.set(p, v) for (p, v) in props.items()]
+
+        return protonode
+
+    def loadNode(self, node):
+        protonode = self.protonodes.get(node.ndef)
+        if protonode is None:
+            protonode = ProtoNode(self, node.buid, node.form, node.ndef[1], node)
+            self.protonodes[node.ndef] = protonode
+        return protonode
+
+    async def _initProtoNode(self, form, norm, norminfo):
+
+        ndef = (form.name, norm)
+
+        protonode = self.protonodes.get(ndef)
+        if protonode is not None:
+            return protonode
+
+        buid = s_common.buid(ndef)
+        node = await self.snap.getNodeByBuid(buid)
+
+        protonode = ProtoNode(self, buid, form, norm, node)
+
+        self.protonodes[ndef] = protonode
+
+        subs = norminfo.get('subs')
```

Comment:
Might not be worth dealing with yet, but we would need to handle `adds` here as well for array forms